spotipy/client.py

```diff
@@ -458,11 +519,74 @@ def search(self, q, limit=10, offset=0, type="track", market=None):
                 - type - the type of item to return. One of 'artist', 'album',
                          'track', 'playlist', 'show', or 'episode'
                 - market - An ISO 3166-1 alpha-2 country code or the string
-                           from_token.
-        """
-        return self._get(
-            "search", q=q, limit=limit, offset=offset, type=type, market=market
-        )
+                           from_token. Can supply list of markets. Pass "ALL" for to search all country codes.
+                - n - return as soon as n options are found
+        """
+        # if string passed and equals "ALL"
+        if (isinstance(market, str) and market.upper() == "ALL"):
+            warnings.warn(
+                "Searching all markets is poorly performing.",
+                UserWarning,
+            )
+            results = {
+                type + 's': {
+                    'href': [],
+                    'items': [],
+                    'limit': limit,
+                    'next': None,
+                    'offset': 0,
+                    'previous': None,
+                    'total': 0
+                }
+            }
+            for country in self.country_codes:
+                result = self._get(
+                    "search", q=q, limit=limit, offset=offset, type=type, market=country
+                )
+                results[type + 's']['href'].append(result[type + 's']['href'])
+                results[type + 's']['items'] += result[type + 's']['items']
+                results[type + 's']['total'] += result[type + 's']['total']
+                if n and len(results[type + 's']['items']) >= n:
+                    return results
+
+            # raise warning
+            return results
+
+        # if list is passed
+        elif isinstance(market, list):
+            warnings.warn(
+                "Searching multiple markets is poorly performing.",
+                UserWarning,
+            )
+            results = {
+                type + 's': {
+                    'href': [],
+                    'items': [],
+                    'limit': limit,
+                    'next': None,
+                    'offset': 0,
+                    'previous': None,
+                    'total': 0
+                }
+            }
+            for country in market:
+                result = self._get(
+                    "search", q=q, limit=limit, offset=offset, type=type, market=country
+                )
+                results[type + 's']['href'].append(result[type + 's']['href'])
+                results[type + 's']['items'] += result[type + 's']['items']
+                results[type + 's']['total'] += result[type + 's']['total']
+                if n and len(results[type + 's']['items']) >= n:
+                    return results
+
+            # raise warning
+            return results
```

Comment:
This part is almost identical to the part above apart from the list of markets, any chance we can move it to a function?

Reply:
Good idea. I can make an internal `self._search_multiple_markets` function to handle both cases